core/domain/user_jobs_one_off.py

```diff
@@ -470,3 +471,142 @@ def map(user_settings_model):
+                item.completed_explorations)
+            completed_activities_model.put()
+            yield ('Regenerated Missing CompletedActivitiesModel', item.id)
+
+        if completed_activities_model.deleted:
+            completed_activities_model.deleted = False
+            completed_activities_model.put()
+            yield (
+                'Reverted delete status for CompletedActivitiesModel',
+                item.id)
+
+        col_model = collection_models.CollectionModel.get_by_id(
+            item.collection_id)
+
+        collection_node_ids = [
+            node['exploration_id'] for node in (
+                col_model.collection_contents['nodes'])]
+        exp_ids_to_remove = [
+            exp_id
+            for exp_id in item.completed_explorations if exp_id not in (
+                collection_node_ids)]
+
+        if exp_ids_to_remove:
+            item.completed_explorations = [
+                exp_id for exp_id in item.completed_explorations
+                if exp_id not in exp_ids_to_remove]
+            item.put()
+            yield (
+                'Invalid Exploration IDs cleaned from '
+                'CollectionProgressModel',
+                'Model id: %s, Collection id: %s, Removed exploration '
+                'ids: %s' % (
+                    item.id, item.collection_id, exp_ids_to_remove))
+
+    @staticmethod
+    def reduce(key, values):
+        values.sort()
+        yield (key, values)
+
+
+class CleanUpUserContributionsModelOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """Job that cleans up UserContributionsModel by removing deleted
+    explorations from user contribution.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.UserContributionsModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        fetched_created_exploration_model_instances = (
+            datastore_services.fetch_multiple_entities_by_ids_and_models(
+                [('ExplorationModel', item.created_exploration_ids)]))[0]
+
+        exp_ids_removed = []
+        for exp_id, exp_instance in list(python_utils.ZIP(
+                copy.deepcopy(item.created_exploration_ids),
+                fetched_created_exploration_model_instances)):
+            if exp_instance is None or exp_instance.deleted:
+                exp_ids_removed.append(exp_id)
+                item.created_exploration_ids.remove(exp_id)
+
+        fetched_edited_exploration_model_instances = (
+            datastore_services.fetch_multiple_entities_by_ids_and_models(
+                [('ExplorationModel', item.edited_exploration_ids)]))[0]
+
+        for exp_id, exp_instance in list(python_utils.ZIP(
+                copy.deepcopy(item.edited_exploration_ids),
+                fetched_edited_exploration_model_instances)):
+            if exp_instance is None or exp_instance.deleted:
+                exp_ids_removed.append(exp_id)
```

Comment:
Could you point to the line of code that does this removal when an exploration is deleted in the future? (Again, with the intent of making sure that no new errors occur.)

Reply:
There is no such code - the way such dependencies are currently handled is via clean up jobs (there is a similar job to clean up user subscription model) since we do not store related model ids. This requires an update to the framework to ensure that external model ids are stored which is being stored here: https://github.com/oppia/oppia/issues/10809